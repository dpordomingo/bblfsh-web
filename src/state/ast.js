import NodeIndex from '../services/NodeIndex';

export const NODE_EXPAND = 'bblfsh/ast/NODE_EXPAND';
export const NODE_TOGGLE = 'bblfsh/ast/NODE_TOGGLE';
export const NODE_HIGHLIGHT = 'bblfsh/ast/NODE_HIGHLIGHT';
export const NODE_UNHIGHLIGHT = 'bblfsh/ast/NODE_UNHIGHLIGHT';

export const nodeReducer = (state, action) => {
  switch (action.type) {
    case NODE_EXPAND:
      return {
        ...state,
        expanded: true,
      };
    case NODE_TOGGLE:
      return {
        ...state,
        expanded: !state.expanded,
      };
    case NODE_HIGHLIGHT:
      return {
        ...state,
        highlighted: true,
      };
    case NODE_UNHIGHLIGHT:
      return {
        ...state,
        highlighted: false,
      };
    default:
      return state;
  }
};

export const DEFAULT_EXPAND_LEVEL = 2;
export const convertTree = uast => {
  const posIndex = new NodeIndex();
  const tree = {};
  let id = 0;

  function convertNode(uast, level, parentId) {
    const curId = ++id;

    const node = {
      ...uast,
      id: curId,
      expanded: level < DEFAULT_EXPAND_LEVEL,
      higlighted: false,
      parentId,
    };
    if (node.Children) {
      node.Children = node.Children.map(child =>
        convertNode(child, level + 1, curId)
      );
    }

    posIndex.add({
      id: node.id,
      start: node.StartPosition,
      end: node.EndPosition,
    });
    tree[curId] = node;

    return curId;
  }

  convertNode(uast, 0);
  return { tree, posIndex };
};

export const expand = nodeId => ({
  type: NODE_EXPAND,
  nodeId,
});

export const nodeHighlight = nodeId => ({
  type: NODE_HIGHLIGHT,
  nodeId,
});

export const nodeUnhighlight = nodeId => ({
  type: NODE_UNHIGHLIGHT,
  nodeId,
});
