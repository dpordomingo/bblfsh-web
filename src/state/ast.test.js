import { convertTree } from './ast';

it('convertTree', () => {
  const uast = {
    InternalType: 'Root',
    Children: [
      {
        InternalType: 'Child1',
        Children: [
          {
            InternalType: 'Child11',
            Children: [{ InternalType: 'Child111' }],
          },
          {
            InternalType: 'Child12',
          },
        ],
      },
      {
        InternalType: 'Child2',
        Children: [
          {
            InternalType: 'Child21',
          },
        ],
      },
    ],
  };

  const tree = convertTree(uast);
  expect(tree).toMatchSnapshot();
});
