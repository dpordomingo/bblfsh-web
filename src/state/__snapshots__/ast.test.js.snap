// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertTree 1`] = `
Object {
  "1": Object {
    "Children": Array [
      2,
      6,
    ],
    "InternalType": "Root",
    "expanded": true,
    "higlighted": false,
    "id": 1,
  },
  "2": Object {
    "Children": Array [
      3,
      5,
    ],
    "InternalType": "Child1",
    "expanded": true,
    "higlighted": false,
    "id": 2,
  },
  "3": Object {
    "Children": Array [
      4,
    ],
    "InternalType": "Child11",
    "expanded": false,
    "higlighted": false,
    "id": 3,
  },
  "4": Object {
    "InternalType": "Child111",
    "expanded": false,
    "higlighted": false,
    "id": 4,
  },
  "5": Object {
    "InternalType": "Child12",
    "expanded": false,
    "higlighted": false,
    "id": 5,
  },
  "6": Object {
    "Children": Array [
      7,
    ],
    "InternalType": "Child2",
    "expanded": true,
    "higlighted": false,
    "id": 6,
  },
  "7": Object {
    "InternalType": "Child21",
    "expanded": false,
    "higlighted": false,
    "id": 7,
  },
}
`;
